### ISSUE TITLE
TestContainer fluent API returning 'this' type

### ISSUE THREAD

**Issue by r3mi** [testcontainers/testcontainers-node#51]
Hello,
thanks for this useful library.
I am trying to extend (subclass) `GenericContainer` but the current typing for the `withXxx` method gets in the way. To be able to extend the fluent API, these methods should return the actual type of Container class, not the base `TestContainer`.
In other words, simply return a [`this type`](https://www.typescriptlang.org/docs/handbook/advanced-types.html#polymorphic-this-types) from the methods e.g.
```ts
export interface TestContainer {
  withEnv(key: EnvKey, value: EnvValue): this;
  withCmd(cmd: Command[]): this;
  // ...
}
```

**Comment by cristianrgreco** [comment]
Hi @r3mi 

Could you show me an example of where it's getting in the way so that I can better understand the problem? I'd guess that the base interface of `TestContainer` is fine as `GenericContainer` implements it anyways. So you'd end up with `CustomGenericContainer -> GenericContainer -> TestContainer`.

**Comment by r3mi** [comment]
Hi @cristianrgreco 
sorry my issue was a bit short. Here are simplified examples:
```ts
class CustomGenericContainer extends GenericContainer {
  ...
  withSomething(): this
  ...
}
const container = await new CustomGenericContainer()
    .withStartupTimeout(new Duration(20, TemporalUnit.SECONDS))
    .withSomething() // <-- HERE can't compile because we are back on a TestContainer
    .start()
```
Another example:
```ts
class CustomGenericContainer extends GenericContainer {
  ...
  start(): Promise<CustomStartedTestContainer>
  ...
}
const container = await new CustomGenericContainer()
    .withStartupTimeout(new Duration(20, TemporalUnit.SECONDS))
    .start()
container.customMethod() // <-- HERE can't compile, container is not expected "CustomStartedTestContainer"
```
Both examples would be fine if `withStartupTimeout` was declared to return '`this` type' instead of `TestContainer`.

**Comment by cristianrgreco** [comment]
@r3mi Please review https://github.com/testcontainers/testcontainers-node/pull/52

**Comment by cristianrgreco** [comment]
@r3mi Available in 2.4.0. Could you have a try and let me know?

**Comment by r3mi** [comment]
works fine, thank you !
